File: solutions/prospective/algorithms/CLRS/insertion_sort.py
"""
Insertion Sort

Time Complexity: O(n^2)
Space Complexity: O(1)

Builds the sorted array one item at a time. Efficient for small data sets.
"""
=== MEMO START ===
def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i
        while j > 0 and arr[j-1] > key:
            arr[j] = arr[j-1]
            j -= 1
        arr[j] = key
    return arr

File: solutions/prospective/algorithms/CLRS/c/insertion_sort.c
/*
Insertion Sort

Time Complexity: O(n^2)
Space Complexity: O(1)

Builds the sorted array one item at a time. Efficient for small data sets.
*/
// === MEMO START ===
void insertion_sort(int *arr, int n) {
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i;
        while (j > 0 && arr[j - 1] > key) {
            arr[j] = arr[j - 1];
            j--;
        }
        arr[j] = key;
    }
}

